Add copy and cut clipboard support, drag and drop and minor change

SioGabx opened the pull request:
### Changelog :
* Copy and Cut to clipboard support (Before, we were getting in the clipboard an empty string)
* If the element is valid when leaving the field, it is added to the list even if it does not end with a ";" (if the isnt in the list, it require SupportsNewObjectCreation to be true on ILookUpContract) -> suggestion of this issue https://github.com/nilayjoshi89/BlackPearl/issues/24#issue-1307165068
* It's now possible to add multiple ItemSeparator at the same time using AdditionalItemSeparators property, for exemple ';', ',', '/' and more...
* Drag and drop support (between MultiSelectCombobox and other text input) 
* Element can be reorder now

PS: I deleted the ARCHIVE folder which dated from 2 years ago ([can be always browse here](https://github.com/nilayjoshi89/BlackPearl/tree/59f3f506e9f1f0ddc9f1ee6caccf293fd400632d/Source/Archive) )

### Copy exemple gif :
<img src="https://user-images.githubusercontent.com/41244237/227998320-debff591-bd68-4547-b991-01ac0143271c.gif" alt="copy_paste-min" width="500" />


### Drag and drop exemples :
<img src="https://user-images.githubusercontent.com/41244237/227998327-5bbf114d-1b65-42fd-be75-1e7927fed704.gif" alt="drag_drop-min2" width="500" />
<img src="https://user-images.githubusercontent.com/41244237/227998329-01295e9c-d611-4ea8-ac89-e3cce641a7b0.gif" alt="drag_drop-min" width="500" />

### New entry in the dropdown" that states that it will create a new item
<img src="https://user-images.githubusercontent.com/41244237/229314191-21243945-f64f-459f-9be8-6384ac92aa6d.png" alt="'new entry in the dropdown' that states that it will create a new item" width="500" />

SioGabx committed:
Added copy to clipboard support and add item if value is compactible

SioGabx committed:
Update readme

SioGabx committed:
Support cut command

SioGabx committed:
Add property to add multiple ItemSeparator at the same time

SioGabx committed:
Update readme and minor change

SioGabx committed:
support for drag and drop
support ordering

SioGabx committed:
drag and drop bug fix and remove useless using

SioGabx committed:
fix string input bug

SioGabx committed:
Added gif to docs folder

SioGabx committed:
Update csproj for nuggets package release

SioGabx committed:
update for nugget package release

SioGabx committed:
Fix null exception on drag and drop

nilayjoshi89 commented on Source/CoreLibrary/Controls/MultiSelectCombobox/BasicStructure.cs:
```diff
@@ -45,6 +45,7 @@ private RichTextBox RichTextBoxElement
                     richTextBoxElement.TextChanged -= RichTextBoxElement_TextChanged;
                     richTextBoxElement.SizeChanged -= RichTextBoxElement_SizeChanged;
                     DataObject.RemovePastingHandler(richTextBoxElement, PasteHandler);
+                    richTextBoxElement.RemoveHandler(CommandManager.PreviewExecutedEvent, new ExecutedRoutedEventHandler(SetClipboardTextWithCommandCancelled));
```
Can you please unregister newly added event handlers here? (Line 69)

DataObject.AddCopyingHandler(richTextBoxElement, OnSelectionStartDrag);
                    richTextBoxElement.DragEnter += OnDragEnter;
                    richTextBoxElement.Drop += OnDragDrop;

nilayjoshi89 commented on Source/CoreLibrary/Controls/MultiSelectCombobox/BasicStructure.cs:
```diff
@@ -235,7 +252,7 @@ private static void DisplayMemberPathChanged(DependencyObject d, DependencyPrope
             handlerType: typeof(SelectionChangedEventHandler), ownerType:
             typeof(MultiSelectCombobox));
-        public event SelectionChangedEventHandler SelectionChanged
+        public event System.EventHandler<SelectionChangedEventArgs> SelectionChanged
```
This will be a breaking change. Let's keep this as is. Any specific reason for this change?

nilayjoshi89 commented on Source/CoreLibrary/Controls/MultiSelectCombobox/EntensionMethods.cs:
```diff
@@ -157,8 +157,10 @@ private static void SelectMultipleItem(this ListBox suggestionList, int delta)
+        [Flags]
         internal enum SuggestionCleanupOperation
         {
+            None = 0,
```
I deliberately skipped the default flag to let the code break if not set properly. Moreover, I don't see 'None' being used. Thoughts?

nilayjoshi89 commented on Source/CoreLibrary/Controls/MultiSelectCombobox/Implementation.cs:
```diff
@@ -17,60 +15,70 @@ public sealed partial class MultiSelectCombobox
         private bool isHandlerRegistered = true;
         private readonly object handlerLock = new object();
+
+        private static RichTextBox DragRichTextBoxValueSource;
```
Static variables in Control are very hard to manage and keep clean. Why can't we rely on 'sender' object?

nilayjoshi89 commented on Source/CoreLibrary/Controls/MultiSelectCombobox/Implementation.cs:
```diff
@@ -17,60 +15,70 @@ public sealed partial class MultiSelectCombobox
         private readonly object handlerLock = new object();
+
+        private static RichTextBox DragRichTextBoxValueSource;
+        private string LastRichTextBoxValue;
```
I don't understand the purpose of this variable. Why do we need to keep track of what was drag-dropped last? There's already duplicate entry handling within control.

SioGabx committed:
Bug fix :
 - wrong order when draging or pasting multiple item
 - item was sometimes disapear on drag and drop

SioGabx committed:
add a "new entry in the dropdown" that states that it will create a new item like suggested in https://github.com/nilayjoshi89/BlackPearl/issues/24
add delay search https://github.com/nilayjoshi89/BlackPearl/issues/15
add more exemple for testing

SioGabx committed:
Changed values to be able to merge with nilayjoshi89/BlackPearl
 - Set version to 2.0.4
 - remove useless usings

SioGabx committed:
indent

SioGabx committed:
Removed Delay search, cause drag and drop issues. Need to be fix (maybe delayed on key down and not textchanged

SioGabx committed:
minor fix

SioGabx committed:
fixed delayed search

SioGabx committed:
drag and drop bug fix from https://github.com/SioGabx/BlackPearl/commit/425da54ab5b8c5fc846771e829770114c84b5eda

SioGabx committed:
added some comments

SioGabx committed:
Rename CustomRichTextBox to MultiselectInternalRichTextBox

SioGabx committed:
fix blink effect on drag drop at the same position

SioGabx committed:
Bug fix : Cannot drag and drop in another Richtextbox

SioGabx committed:
updated code to be working with objects

DSPaul commented:
Hey there, Its great to see the issue I made months ago has been addressed but I am a bit confused as to the state of this PR. I saw you made a release (v2.1.4) over in your fork, is that release production ready? And if it is, what is stopping this draft from becoming a proper PR ready to merge? 

I was even more confused when I saw that the owner of this repo is now committing to a branch in your fork and has made a PR against the master of that fork which seems completely backwards. Why is he not committing his changes to his own repo? 

To be clear, I'm not telling anyone here what to do, I'm just really confused and eager to finally use the fix to my issue that I have been waiting on for close to a year and that has seemingly just been sitting here, ready, for over a month.

SioGabx commented:
> Hey there, Its great to see the issue I made months ago has been addressed but I am a bit confused as to the state of this PR. I saw you made a release (v2.1.4) over in your fork, is that release production ready? And if it is, what is stopping this draft from becoming a proper PR ready to merge?
> 
> I was even more confused when I saw that the owner of this repo is now committing to a branch in your fork and has made a PR against the master of that fork which seems completely backwards. Why is he not committing his changes to his own repo?
> 
> To be clear, I'm not telling anyone here what to do, I'm just really confused and eager to finally use the fix to my issue that I have been waiting on for close to a year and that has seemingly just been sitting here, ready, for over a month.

Hey !
We are still currently working on it, it should be good soon I think (I understand your confusion, here is some explanation on what is happening).
First of all, my "release" in my fork is not ready for production, I had done it because I had not heard back at that time from @nilayjoshi89. I just put it as a pre-release.

Since then, we have exchanged together by email and are doing code reviews on ways to improve the pull request (in a cleaner way than what I had done and making sure that there are no bugs).
As we had a lot of things to review, we chose to edit the code directly in my branch.

As of this writing, the most up-to-date branch is mine (https://github.com/SioGabx/BlackPearl/tree/dev2). I am waiting for feedback from @nilayjoshi89.

For the moment, among your two suggestions that I had integrated, we have only implemented the fact that If the element is valid when leaving the field, it is added to the list even if it does not end with a ";"
The second suggestion ("new entry in the dropdown" that states that it will create a new item) will be done later when we finish the first part.

nilayjoshi89 committed:
Keeping template part RichTextBox, same as before, for consistency and backward compatibility.

nilayjoshi89 committed:
Cleaning Generic.xaml

DSPaul commented:
I see, thank you for clearing that up, looking forward to it!

nilayjoshi89 committed:
Code cleanup

SioGabx committed:
Fix issues where item was vanishes on drag and drop but SelectedItem have still retain those items

nilayjoshi89 committed:
Minor changes

SioGabx committed:
Fix refactoring issues :
- if the data is an object, it was never returned
- if its on the same richTextBox and the drag and drop position is the same as actual, then we do not perform OnDragDrop. (added comments)

SioGabx committed:
- Added function RemoveDiacritics to the DefaultLookUpContract to ignore diacritics on text compare ex: "héllo" now match "hello".
- Added somes dummy data with names with diacritics for testing purpose.
- Fix issues : item disapear when pasting existing item inside the element.
- Fix crash when we were doing multiples copy of a item

SioGabx committed:
Bug fix : if the user paste and has a item inside the selection, then we need to replace it with the pasted content.

SioGabx committed:
refactoring previous bug fixes

SioGabx committed:
minor fix

SioGabx committed:
Created View/ViewModels for the Demo application

SioGabx committed:
Merge pull request #2 from SioGabx/devs/nilay/dev2

Devs/nilay/dev2
